### PR TITLE
Make unmarshal faster

### PR DESCRIPTION
**Pull request by rooftoofwoof**
Unmarshalling into structs became theoretically quadratic slow because of a recent refactor #95 . This PR fixes it by initializing features in a scoped manner using a trie. Previously, to initialize each feature, we were looping through every field in every namespace in a feature's path. Now we simply do a one pass over the feature graph to initialize all features.

### CODE REVIEW

File: init_features.go [hunk 1]
```diff
@@ -3,57 +3,101 @@ package chalk
...
+
+	firstPart := fqnParts[0]
+	if s.Children == nil {
+		s.Children = map[string]*scopeTrie{}
+	}
+
+	kid := s.Children[firstPart]
+	if kid == nil {
+		s.Children[firstPart] = &scopeTrie{}
+	}
```
Comment (rooftoofwoof) [inline init_features.go:39]: use `ok` and inline

File: init_features.go [hunk 2]
```diff
@@ -3,57 +3,101 @@ package chalk
...
+
+	s.Children[firstPart].add(fqnParts[1:])
+}
+
+type featureInitializer struct {
+	fieldsMap map[string][]reflect.Value
+	isScoped  bool
```
Comment (rooftoofwoof) [inline init_features.go:46]: maybe just use the scope param to determine whether in scope

File: init_features.go [hunk 3]
```diff
@@ -159,68 +207,98 @@ func initFeatures(
...
-				pattern, err := regexp.Compile(fmt.Sprintf(`^%s__\d+__$`, resolvedName))
-				if err != nil {
-					return nil, errors.Wrap(err, "error compiling regex to match windowed feature names")
-				}
-				if !pattern.Match([]byte(getFqnRoot(targetFqn))) {
-					// Not any bucket feature in this windowed feature class
-					continue
-				}
-			}
+			// FIXME: Should be removeable
```
Comment (rooftoofwoof) [inline init_features.go:210]: remove

File: init_features.go [hunk 4]
```diff
@@ -159,68 +207,98 @@ func initFeatures(
...
-			if targetFqn == "" {
-				// Initializing all features, always
-				// make a new map.
-				f.Set(reflect.MakeMap(f.Type()))
-				windows := fm.Meta.Tag.Get("windows")
-				for _, tag := range strings.Split(windows, ",") {
-					seconds, parseErr := internal.ParseBucketDuration(tag)
-					if parseErr != nil {
-						return nil, fmt.Errorf("error parsing bucket duration: %s", parseErr)
+			//if fi.isScoped {
```
Comment (rooftoofwoof) [inline init_features.go:237]: remove

File: init_features.go [hunk 5]
```diff
@@ -3,57 +3,79 @@ package chalk
...
 	"strings"
 	"time"
 )
 
 func InitFeatures[T any](t *T) error {
 	structValue := reflect.ValueOf(t).Elem()
-	_, err := initFeatures(structValue, "", make(map[string]bool), "")
-	return err
+	initializer := NewFeatureInitializer()
+	return initializer.initFeatures(structValue, "", make(map[string]bool), nil)
```
Comment (rooftoofwoof) [inline init_features.go:15]: inline

File: init_features.go [hunk 6]
```diff
@@ -3,57 +3,79 @@ package chalk
...
 }
 
-func initFeatureSingle(structValue reflect.Value, fqn string) ([]reflect.Value, error) {
-	parts := strings.Split(fqn, ".")
-	if len(parts) < 2 {
-		return nil, fmt.Errorf("feature fqn should have at least two parts, found: '%s'", fqn)
+type scopeTrie struct {
+	Children map[string]*scopeTrie
```
Comment (rooftoofwoof) [inline init_features.go:19]: private

File: init_features.go [hunk 7]
```diff
@@ -159,68 +185,54 @@ func initFeatures(
...
-					if parseErr != nil {
-						return nil, fmt.Errorf("error parsing bucket duration: %s", parseErr)
+			windows := fm.Meta.Tag.Get("windows")
+			for _, tag := range strings.Split(windows, ",") {
+				seconds, err := internal.ParseBucketDuration(tag)
+				if err != nil {
+					return errors.Wrap(err, "error parsing bucket duration: %s")
+				}
+				updatedResolvedName := fmt.Sprintf("%s__%d__", resolvedName, seconds)
+				windowFqn := fmt.Sprintf("%s.%s", cumulativeFqn, updatedResolvedName)
```
Comment (rooftoofwoof) [inline init_features.go:204]: bucket fqn

File: unmarshal_test.go [hunk 8]
```diff
@@ -874,3 +874,113 @@ func TestUnmarshalQueryBulkListOfPrimitives(t *testing.T) {
...
+	singleFeatures := &allTypes{}
+	singleAvg := benchmarkUnmarshal(t, singleData, singleFeatures)
+	assert.Equal(t, int64(123), *singleFeatures.Int)
+
+	lenDelta := len(multiData) - len(singleData)
+	delta := multiAvg - singleAvg
+	deltaPerExtraItem := float64(delta) / float64(lenDelta)
+	singleItemDuration := float64(singleAvg)
+	multiplier := deltaPerExtraItem / singleItemDuration
+	assert.True(
```
Comment (rooftoofwoof) [inline unmarshal_test.go:980]: log

File: init_features.go [hunk 9]
```diff
@@ -3,57 +3,78 @@ package chalk
 import (
 	"fmt"
 	"github.com/chalk-ai/chalk-go/internal"
-	"github.com/pkg/errors"
+	"github.com/cockroachdb/errors"
```
Comment (Nathan-Fenner) [inline init_features.go:6]: Is this change intentional?

File: init_features.go [hunk 10]
```diff
@@ -159,68 +184,54 @@ func initFeatures(
...
+					if f.IsNil() {
+						f.Set(reflect.MakeMap(f.Type()))
 					}
+					fi.fieldsMap[bucketFqn] = append(fi.fieldsMap[bucketFqn], f)
+				} else {
+					if f.IsNil() {
+						f.Set(reflect.MakeMap(f.Type()))
+					}
+					feature := Feature{Fqn: bucketFqn}
+					ptrInDisguiseToFeature := reflect.NewAt(mapValueType.Elem(), reflect.ValueOf(&feature).UnsafePointer())
```
Comment (Nathan-Fenner) [inline init_features.go:219]: This doesn't seem like a safe usage of `UnsafePointer` - I think this should just be

```go
reflect.ValueOf(&feature)
```

Comment (Nathan-Fenner) [reply init_features.go:219]: Discussed offline - this seems wrong but isn't new to this PR so it's not a blocker

**Review by Nathan-Fenner — APPROVED** [review]

File: init_features.go (hunk 9 above)
Comment (rooftoofwoof) [reply init_features.go:6]: Yes - I'm trying to phase out the usage of pkg/errors in favor of the cockroach one

File: init_features.go (hunk 10 above)
Comment (rooftoofwoof) [reply init_features.go:219]: ticketed here https://linear.app/chalk-ai/issue/CHA-3991/re-evaluate-chalk-go-deser-implementation-via-unsafepointer
